src/Workspaces/Core/Portable/Remote/RemoteHostSessions.cs

```diff
@@ -0,0 +1,280 @@
+            _session.InvokeAsync(targetName, arguments);
+        public Task<T> InvokeAsync<T>(string targetName, params object[] arguments) =>
+            _session.InvokeAsync<T>(targetName, arguments);
+        public Task InvokeAsync(string targetName, IEnumerable<object> arguments, Func<Stream, CancellationToken, Task> funcWithDirectStreamAsync) =>
+            _session.InvokeAsync(targetName, arguments, funcWithDirectStreamAsync);
+        public Task<T> InvokeAsync<T>(string targetName, IEnumerable<object> arguments, Func<Stream, CancellationToken, Task<T>> funcWithDirectStreamAsync) =>
+            _session.InvokeAsync<T>(targetName, arguments, funcWithDirectStreamAsync);
+    }
+
+    /// <summary>
+    /// Helper type of common session case
+    /// 
+    /// RemoteHostClient.Session that lives for multiple solutions
+    /// </summary>
+    internal sealed class KeepAliveSession
+    {
+        private readonly SemaphoreSlim _gate;
+        private readonly IRemoteHostClientService _remoteHostClientService;
+        private readonly CancellationToken _cancellationToken;
+
+        private readonly string _serviceName;
+        private readonly object _callbackTarget;
+
+        private RemoteHostClient _client;
+        private RemoteHostClient.Session _session;
+
+        public KeepAliveSession(RemoteHostClient client, RemoteHostClient.Session session, string serviceName, object callbackTarget, CancellationToken cancellationToken)
+        {
+            _gate = new SemaphoreSlim(initialCount: 1);
+            _remoteHostClientService = client.Workspace.Services.GetService<IRemoteHostClientService>();
+            _cancellationToken = cancellationToken;
+
+            _serviceName = serviceName;
+            _callbackTarget = callbackTarget;
+
+            Initialize_NoLock(client, session);
+        }
+
+        public void Shutdown()
+        {
+            using (_gate.DisposableWait(_cancellationToken))
+            {
+                if (_client != null)
+                {
+                    _client.ConnectionChanged -= OnConnectionChanged;
+                }
+
+                _session?.Dispose();
+
+                _client = null;
+                _session = null;
+            }
+        }
+
+        public async Task<bool> TryInvokeAsync(string targetName, params object[] arguments)
+        {
+            using (await _gate.DisposableWaitAsync(_cancellationToken).ConfigureAwait(false))
+            {
+                var session = await TryGetSession_NoLockAsync().ConfigureAwait(false);
+                if (session == null)
+                {
+                    return false;
+                }
+
+                await session.InvokeAsync(targetName, arguments).ConfigureAwait(false);
+                return true;
+            }
+        }
+
+        public async Task<(bool success, T result)> TryInvokeAsync<T>(string targetName, params object[] arguments)
```

Comment:
all code is little bit different so not sure. are you asking using lambda to change those parts? most of time, that makes code harder to udnerstand though since code are spread around